<html >
<head>
<meta http-equiv="Content-Language" content="en-us">
<meta http-equiv="Content-Type" content="text/html;">
<title>InterVar</title>
<meta name="Intervar" content="Intervar">
</head>

<body bgcolor='white'>
<table border="0" width="100%" id="table1">
	<tr>
    <tr>
        <td class="style2"><img border="0" src="mcgill.jpg"  height="100"></td>

        <td align=left class="style3"><i><b><font face="Times New Roman" size="7" color="#0000FF">
            InterVar</font></b></i><em><font face="Times New Roman" color="#0000ff" size="6"><strong><br>
<!--
                        T</strong></font><strong><font face="Times New Roman" size="6">ype &nbsp;<span
                class="style4">2</span></font><font face="Times New Roman" color="#0000ff" size="6">
                        &nbsp;D</font><font face="Times New Roman" size="6">iabetes</font><font face="Times New Roman" color="#0000ff" size="6">
            &nbsp;R</font><font face="Times New Roman" size="6">isk</font>
            <font face="Times New Roman" color="#0000ff" size="6">
                        P</font><font face="Times New Roman" size="6">rediction</font><font face="Times New Roman" color="#0000ff" size="6">
            <br /> -->
            (<i><span lang="EN-US" style="font-size:15.0pt;
mso-bidi-font-size:24.0pt;line-height:115%;font-family:&quot;Calibri&quot;,&quot;sans-serif&quot;;
mso-ascii-theme-font:minor-latin;mso-fareast-font-family:ו;mso-fareast-theme-font:
minor-fareast;mso-hansi-theme-font:minor-latin;mso-bidi-font-family:&quot;Times New Roman&quot;;
mso-bidi-theme-font:minor-bidi;mso-ansi-language:EN-US;mso-fareast-language:
ZH-CN;mso-bidi-language:AR-SA">for Inter from ACMG2015</span></i>)</strong></em></td>         <td class="style1"><img border="0" src="uthsc.jpg"  height="100"  style="width: 150px"></td>
    </tr>
</table>



<td width="75%" valign="top">

 <hr>
<form enctype="multipart/form-data" action="mlr.php" method="post" bgcolor='#ddddee'>

<hr>
<p class="pvs">
 <input type="checkbox" name="pvs1" value="1"> PVS1: null variant (nonsense, frameshift, canonical +- 2 splice sites, initiation codon, single or multiexon deletion) in a gene where LOF is a known mechanism of disease<br>
</p>
<p class="pvs">
<input type="checkbox" name="ps1" value="1"> PS1: Same amino acid change as a previously established pathogenic variant regardless of nucleotide change<br>
<input type="checkbox" name="ps2" value="1"> PS2: De novo (both maternity and paternity confirmed) in a patient with the disease and no family history<br>
<input type="checkbox" name="ps3" value="1"> PS3: Well-established in vitro or in vivo functional studies supportive of a damaging effect on the gene or gene product<br>
<input type="checkbox" name="ps4" value="1"> PS4: The prevalence of the variant in affected individuals is significantly increased compared with the prevalence in controls<br>
<!-- <input type="checkbox" name="pp1" value="1"> PP1 (Strong evidence) Cosegregation with disease in multiple affected family members in a gene definitively known to cause the disease<br> -->
</p>
<p class="pm">
<input type="checkbox" name="pm1" value="1"> PM1: Located in a mutational hot spot and/or critical and well-established functional domain (e.g., active site of an enzyme) without benign variation<br>
<input type="checkbox" name="pm2" value="1"> PM2: Absent from controls (or at extremely low frequency if recessive) in Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation Consortium<br>
<input type="checkbox" name="pm3" value="1"> PM3: For recessive disorders, detected in trans with a pathogenic variant<br>
<input type="checkbox" name="pm4" value="1"> PM4: Protein length changes as a result of in-frame deletions/insertions in a nonrepeat region or stop-loss variants<br>
<input type="checkbox" name="pm5" value="1"> PM5: Novel missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before<br>
<input type="checkbox" name="pm6" value="1"> PM6: Assumed de novo, but without confirmation of paternity and maternity<br>
<!-- <input type="checkbox" name="pp1_m" value="1"> PP1 (Moderate evidence) Cosegregation with disease in multiple affected family members in a gene definitively known to cause the disease<br> -->
</p>
<p class="pp">
<input type="checkbox" name="pp1" value="1"> PP1: Cosegregation with disease in multiple affected family members in a gene definitively known to cause the disease<br>
<input type="checkbox" name="pp2" value="1"> PP2: Missense variant in a gene that has a low rate of benign missense variation and in which missense variants are a common mechanism of disease<br>
<input type="checkbox" name="pp3" value="1"> PP3: Multiple lines of computational evidence support a deleterious effect on the gene or gene product (conservation, evolutionary, splicing impact, etc.)<br>
<input type="checkbox" name="pp4" value="1"> PP4: Patient's phenotype or family history is highly specific for a disease with a single genetic etiology<br>
<input type="checkbox" name="pp5" value="1"> PP5: Reputable source recently reports variant as pathogenic, but the evidence is not available to the laboratory to perform an independent evaluation
</p>
<p class="bs">
<input type="checkbox" name="ba1" value="1"> BA1: Allele frequency is >5% in Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation Consortium<br>
</p>
<p class="bs">
<input type="checkbox" name="bs1" value="1"> BS1: Allele frequency is greater than expected for disorder<br>
<input type="checkbox" name="bs2" value="1"> BS2: Observed in a healthy adult individual for a recessive (homozygous), dominant (heterozygous), or X-linked (hemizygous) disorder, with full penetrance expected at an early age<br>
<input type="checkbox" name="bs3" value="1"> BS3: Well-established in vitro or in vivo functional studies show no damaging effect on protein function or splicing<br>
<input type="checkbox" name="bs4" value="1"> BS4: Lack of segregation in affected members of a family<br>
</p>
<p class="bp">
<input type="checkbox" name="bp1" value="1"> BP1: Missense variant in a gene for which primarily truncating variants are known to cause disease<br>
<input type="checkbox" name="bp2" value="1"> BP2: Observed in trans with a pathogenic variant for a fully penetrant dominant gene/disorder or observed in cis with a pathogenic variant in any inheritance pattern<br>
<input type="checkbox" name="bp3" value="1"> BP3: In-frame deletions/insertions in a repetitive region without a known function<br>
<input type="checkbox" name="bp4" value="1"> BP4: Multiple lines of computational evidence suggest no impact on gene or gene product (conservation, evolutionary, splicing impact, etc.)<br>
<input type="checkbox" name="bp5" value="1"> BP5: Variant found in a case with an alternate molecular basis for disease<br>
<input type="checkbox" name="bp6" value="1"> BP6: Reputable source recently reports variant as benign, but the evidence is not available to the laboratory to perform an independent evaluation<br>
<input type="checkbox" name="bp7" value="1"> BP7: A synonymous (silent) variant for which splicing prediction algorithms predict no impact to the splice consensus sequence nor the creation of a new splice site AND the nucleotide is not highly conserved<br>
</p>


 <INPUT TYPE="submit" VALUE="submit">
 <INPUT TYPE="reset" VALUE="Clear">
 </DIV>
</p>
 <HR>
 </p>
 </DIV>
</FORM>

</td>
</body>
</html>
